psi4/driver/procrouting/sapt/sapt_mp2_terms.py

```diff
@@ -112,75 +116,113 @@ def df_fdds_dispersion(primary, auxiliary, cache, leg_points=10, leg_lambda=0.3,
 
     # Builds potentials
     W_A = fdds_obj.metric().clone()
-    W_A.axpy(1.0, _compute_fxc(D[0], half_Saux, halfp_Saux, rho_thresh=rho_thresh))
+    W_A.axpy(1.0, _compute_fxc(D[0], half_Saux, halfp_Saux, x_alpha, rho_thresh=rho_thresh))
+    W_A = W_A.to_array()
 
     W_B = fdds_obj.metric().clone()
-    W_B.axpy(1.0, _compute_fxc(D[1], half_Saux, halfp_Saux, rho_thresh=rho_thresh))
+    W_B.axpy(1.0, _compute_fxc(D[1], half_Saux, halfp_Saux, x_alpha, rho_thresh=rho_thresh))
+    W_B = W_B.to_array()
 
     # Nuke the densities
     del D
 
-    metric = fdds_obj.metric()
-    metric_inv = fdds_obj.metric_inv()
+    metric = fdds_obj.metric().clone().to_array()
+    metric_inv = fdds_obj.metric_inv().clone().to_array()
 
     # Integrate
     core.print_out("\n   => Time Integration <= \n\n")
 
     val_pack = ("Omega", "Weight", "Disp20,u", "Disp20", "time [s]")
     core.print_out("% 12s % 12s % 14s % 14s % 10s\n" % val_pack)
-    # print("% 12s % 12s % 14s % 14s % 10s" % val_pack)
     start_time = time.time()
 
     total_uc = 0
     total_c = 0
 
+    # Read R
+    if is_hybrid:
+        R_A = fdds_obj.R_A().to_array()
+        R_B = fdds_obj.R_B().to_array()
+        Rtinv_A = np.linalg.pinv(R_A, rcond=1.e-13).transpose()
+        Rtinv_B = np.linalg.pinv(R_B, rcond=1.e-13).transpose()
+
     for point, weight in zip(*np.polynomial.legendre.leggauss(leg_points)):
 
         omega = leg_lambda * (1.0 - point) / (1.0 + point)
         lambda_scale = ((2.0 * leg_lambda) / (point + 1.0)**2)
 
         # Monomer A
-        X_A = fdds_obj.form_unc_amplitude("A", omega)
+        if is_hybrid:
+            aux_list = fdds_obj.form_aux_matrices("A", omega)
+            aux_length = len(aux_list)
+            for i in range(aux_length):
```

Comment:
`aux_list = [x.to_array() for x in aux_list]`

Reply:
So the main reason I'm introducing the aux_list is to make it more convenient for `to_array()` all of them. If we do `matrix_name_a, b, c, d = fdds_obj.form...` in line 156, is there any quick way to convert all of them to numpy array afterwards?

Reply:
I think DGAS is suggesting that the C++ fn return a map of strings to psi4.core.Matrix objects. Then the fn here has a single return value of the map. That map behaves like a python dict, so then you can `aux_dict = {k: v.to_array() for k, v in aux_dict.items()}` to numpy-ize them. `Process::environment.arrays` is an example of a similar C++ map as Python dict.